regularizepsf/visualize.py

```diff
@@ -0,0 +1,125 @@
+import matplotlib
+import matplotlib.pyplot as plt
+import numpy as np
+
+from regularizepsf.fitter import PatchCollectionABC
+
+
+def visualize_patch_counts(patch_collection: PatchCollectionABC,
+                           ax: matplotlib.axes.Axes = None) -> matplotlib.axes.Axes:
+    """
+    Utility to visualize the number of stars identified within each patch
+
+    Parameters
+    ----------
+    patch_collection : PatchCollectionABC
+        A patch collection, such as that returned by
+        `CoordinatePatchCollection.find_stars_and_average`.
+    ax : matplotlib.axes.Axes
+        An Axes object on which to plot. If not provided, a new Figure will be
+        generated.
+    """
+    if patch_collection.counts is None or not len(patch_collection.counts):
+        raise ValueError("This PatchCollection does not have any counts")
+
+    if ax is None:
+        fig = plt.figure()
+        ax = fig.subplots()
+
+    corners = []
+    counts = []
+    for corner, count in patch_collection.counts.items():
+        corners.append(corner)
+        counts.append(count)
+    # corners has the lower-left corner of each patch
+    corners = np.array(corners)
+    corners += patch_collection.size // 2
+    # Now it has the center of each patch
+
+    ax.set_aspect('equal')
+    m = ax.scatter(corners[:, 1], corners[:, 0], c=counts, s=200)
+    plt.colorbar(m).set_label(
+            "Number of stars found in patch centered on point")
+
+    return ax
+
+
+def visualize_patches(patch_collection: PatchCollectionABC,
```

Comment:
Here's the code I used to visualize a before and after set of patches:
```python
import matplotlib.pyplot as plt
from matplotlib.gridspec import GridSpec
from mpl_toolkits.axes_grid1.inset_locator import inset_axes
import itertools
import matplotlib.pylab as pylab
import matplotlib.colors as colors
params = {'axes.titlesize': 15}
pylab.rcParams.update(params)

vmin = 0.02
vmax = 1.0

fig = plt.figure(constrained_layout=True, figsize=(13, 6))

gs = GridSpec(4, 9, figure=fig, 
              width_ratios=([1]*4) + [0.8] + ([1]*4),
              height_ratios=[1]*4)

uncorrected_axs = dict()
for i, j in itertools.product(range(4), range(4)):
    uncorrected_axs[i, j] = fig.add_subplot(gs[i, j])
    
for i in range(16):
    row, col = i % 4, i // 4
    x = coords[np.array([1, 3, 5, 7])][row]
    y = coords[np.array([1, 3, 5, 7])][col]
    im = uncorrected_axs[3-row, col].imshow(array_corrector[x, y][117:-117, 117:-117], 
                                  origin='lower', cmap=custom, 
                                            norm=colors.PowerNorm(gamma=1/2.2, vmin=vmin, vmax=vmax))
    uncorrected_axs[3-row, col].set_axis_off()
    uncorrected_axs[3-row, col].set_aspect(1)
    
cax = fig.add_axes([0.479, 0.065, 0.012, 0.85])
#fig.colorbar(im, cax=cax, orientation='vertical', label='log$_{10}$(brightness)')
fig.colorbar(im, cax=cax, orientation='vertical', label='brightness')

corrected_axs = dict()
for i, j in itertools.product(range(4), range(4)):
    corrected_axs[i, j] = fig.add_subplot(gs[i, j+5])
for i in range(16):
    row, col = i // 4, i % 4
    x = coords[np.array([1, 3, 5, 7])][row]
    y = coords[np.array([1, 3, 5, 7])][col]
    im = corrected_axs[3-row, col].imshow(output[CoordinateIdentifier(None, x, y)][117:-117, 117:-117], 
                                    origin='lower', cmap=custom, 
                                          norm=colors.PowerNorm(gamma=1/2.2, vmin=vmin, vmax=vmax))
    corrected_axs[3-row, col].set_axis_off()
    corrected_axs[3-row, col].set_aspect(1)

    
plt.figtext(0.23, 0.95, 'Uncorrected', ha='center', va='center', fontsize=15)
plt.figtext(0.23+0.54, 0.95, 'Corrected', ha='center', va='center', fontsize=15)
uncorrected_axs[0, 2].set_title("\n")
plt.show()
fig.savefig("/Users/jhughes/Desktop/punch.png", dpi=300)
```

It would obviously need generalizing if we were to add it.

Reply:
> It would obviously need generalizing if we were to add it.

`fig.savefig("/Users/jhughes/Desktop/punch.png", dpi=300)`

I think we can just require the user to log in as `jhughes`, no need to generalize

Reply:
But I do prefer your patch plots, so I think this code should be the base of what we end up on